src/duv/udp.c

```diff
@@ -1 +1,173 @@
+//  struct sockaddr_storage addr;
+//  if (uv_ip4_addr(host, port, (struct sockaddr_in*)&addr) &&
+//      uv_ip6_addr(host, port, (struct sockaddr_in6*)&addr)) {
+//    duk_error(ctx, DUK_ERR_TYPE_ERROR, "Invalid IP address or port");
+//  }
+//  // TODO : implement flags, store addr6
+//  duv_push_sockaddr(ctx, &addr, sizeof(addr));
+//  //DUK_PUSH_PROP_POINTER("\xff""\xff""data",TP);
+//
+//  return 0;
+//}
+
+duk_ret_t duv_udp_bind(duk_context *ctx) {
+  dschema_check(ctx, (const duv_schema_entry[]) {
+    {"host", duk_is_string},
+    {"port", duk_is_number},
+    {0,0}
+  });
+  uv_udp_t *udp = duv_require_this_handle(ctx, DUV_UDP_MASK);
+  const char *host = duk_get_string(ctx, 1);
+  int port = duk_get_number(ctx, 2),
+      // UV_UDP_REUSEADDR = 4
+      flags = 4;
+  struct sockaddr_storage addr;
+  if (uv_ip4_addr(host, port, (struct sockaddr_in*)&addr) &&
+      uv_ip6_addr(host, port, (struct sockaddr_in6*)&addr)) {
+    duk_error(ctx, DUK_ERR_TYPE_ERROR, "Invalid IP address or port");
+  }
+  // TODO : implement flags
+  duv_check(ctx, uv_udp_bind(udp, (struct sockaddr*)&addr, flags));
+  return 0;
+}
+
+duk_ret_t duv_udp_getsocket(duk_context *ctx) {
+  uv_udp_t *udp = duv_require_this_handle(ctx, DUV_UDP_MASK);
+  struct sockaddr_storage address;
+  int addrlen = sizeof(address);
+  duv_check(ctx, uv_udp_getsockname(udp, (struct sockaddr*)&address, &addrlen));
+  duv_push_sockaddr(ctx, &address, addrlen);
+  return 1;
+}
+
+duk_ret_t duv_udp_getsockname(duk_context *ctx) {
+  uv_udp_t *udp = duv_require_this_handle(ctx, DUV_UDP_MASK);
+  struct sockaddr_storage address;
+  int addrlen = sizeof(address);
+  duv_check(ctx, uv_udp_getsockname(udp, (struct sockaddr*)&address, &addrlen));
+  duv_push_sockaddr(ctx, &address, addrlen);
+  return 1;
+}
+
+duk_ret_t duv_udp_recv_stop(duk_context *ctx) {
+  uv_udp_t *udp = duv_require_this_handle(ctx, DUV_UDP_MASK);
+  duv_check(ctx, uv_udp_recv_stop(udp));
+  return 0;
+}
+
+
+duk_ret_t duv_udp_recv_start(duk_context *ctx) {
+  dschema_check(ctx, (const duv_schema_entry[]) {
+    {"callback", dschema_is_continuation},
+    {0,0}
+  });
+  uv_udp_t *udp = duv_require_this_handle(ctx, DUV_UDP_MASK);
+  duk_put_prop_string(ctx, 0, "\xffon-read");
+  duv_check(ctx, uv_udp_recv_start(udp, duv_on_alloc, duv_on_udp_recv_start));
+  return 0;
+}
+
+duk_ret_t duv_udp_send(duk_context *ctx) {
+  dschema_check(ctx, (const duv_schema_entry[]) {
+    {"host", duk_is_string},
+    {"port", duk_is_number},
+    {"data", dschema_is_data},
+    {"callback", dschema_is_continuation},
+    {0,0}
+  });
+  uv_udp_t *udp = duv_require_this_handle(ctx, DUV_UDP_MASK);
+  const char *host = duk_get_string(ctx, 1);
+  int port = duk_get_number(ctx, 2);
+  //    flags = 0;
+  struct sockaddr_storage addr;
+  if (uv_ip4_addr(host, port, (struct sockaddr_in*)&addr) &&
+      uv_ip6_addr(host, port, (struct sockaddr_in6*)&addr)) {
+    duk_error(ctx, DUK_ERR_TYPE_ERROR, "Invalid IP address or port");
```

Comment:
What if buffer that gets passed to UDP send gets out of scope and collected?
I can't find any head stash reference.. am I missing something here?

On Mon, Aug 15, 2016, 22:52 Tim Caswell <notifications@github.com> wrote:

> In src/duv/udp.c
> <https://github.com/nucleus-js/seaduk/pull/8#discussion_r74822763>:
>
> > +duk_ret_t duv_udp_send(duk_context *ctx) {
> > +  dschema_check(ctx, (const duv_schema_entry[]) {
> > +    {"host", duk_is_string},
> > +    {"port", duk_is_number},
> > +    {"data", dschema_is_data},
> > +    {"callback", dschema_is_continuation},
> > +    {0,0}
> > +  });
> > +  uv_udp_t *udp = duv_require_this_handle(ctx, DUV_UDP_MASK);
> > +  const char *host = duk_get_string(ctx, 1);
> > +  int port = duk_get_number(ctx, 2);
> > +  //    flags = 0;
> > +  struct sockaddr_storage addr;
> > +  if (uv_ip4_addr(host, port, (struct sockaddr_in*)&addr) &&
> > +      uv_ip6_addr(host, port, (struct sockaddr_in6*)&addr)) {
> > +    duk_error(ctx, DUK_ERR_TYPE_ERROR, "Invalid IP address or port");
>
> Doesn't duk_error perform a long jump the same as lua?
>
> I do know that lua had an int return value on it anyway so that return
> could be used to avoid certain compilers from complaining.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nucleus-js/seaduk/pull/8/files/ea255eced5607bd2c5b5d30d0ebc198751b0991b#r74822763>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA7no-6wsQAC7u0pofo3xUVOeg6USUTIks5qgMOZgaJpZM4JkU_7>
> .
>